### DISCUSSION TITLE
Atan in micropython

### DISCUSSION THREAD

**Discussion by Redstonetill** [pybricks/support#1857] [Q&A] [answered]
I currently use the pybricks micropython repo to create my own function inside of the pybricks firmware. But when trying to build the firmware I am getting an error: `<artificial>:(.text.pb_module_tools_get_angle_error.lto_priv.0+0x1e2): undefined reference to atan'`
(same is for `fmod`)

If anyone could help, I would be gratefull

```#include <math.h>```

[...]

```
static mp_obj_t pb_module_tools_get_angle_error(size_t n_args, const mp_obj_t *args) {
    // Stelle sicher, dass 5 Argumente übergeben werden
    if (n_args != 5) {
        mp_raise_ValueError(mp_obj_new_str("Expected 5 arguments", strlen("Expected 5 arguments")));
    }

    // Konvertiere die Argumente in C-Float-Werte
    double startX = mp_obj_get_float(args[0]);
    double startY = mp_obj_get_float(args[1]);
    double pointX = mp_obj_get_float(args[2]);
    double pointY = mp_obj_get_float(args[3]);
    double startAngle = mp_obj_get_float(args[4]);

    double angleToWaypoint = 0.0;

    if (startY == pointY) {angleToWaypoint = startX < pointX ? 0.0 : 180.0;}
    else {
        double dx = pointX - startX;
        double dy = startY - pointY;

        angleToWaypoint = atan(dx / dy) * 180 / (double) M_PI;

        if (dy < 0) {angleToWaypoint += 90;}
        if (dy > 0) {angleToWaypoint -= 90;}
    }

    // --- Get fastes way to turn ---
    double rightDistance = fmod(angleToWaypoint - startAngle + 720, 360.0);
    double leftDistance = fmod(startAngle - angleToWaypoint + 720, 360.0);
    
    if (rightDistance < leftDistance) {return mp_obj_new_float(rightDistance);}
    
    return mp_obj_new_float(-leftDistance);
}

static MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(pb_module_tools_get_angle_Error_obj, 5, 5, pb_module_tools_get_angle_error);
```

**Comment by dlech** [comment]
Which hub are you targeting?

**Reply by Redstonetill** [reply]
Sorry for not stating, I am currently using the Spike Prime Hub

**Comment by dlech** [comment, marked as answer]
Ah, I see the problem now. Pybricks hubs only use single precision floating point, not double

For portability,

```
double startX = mp_obj_get_float(args[0]);
```

should be

```
mp_float_t startX = mp_obj_get_float(args[0]);
```

or to force single-precision, you can do

```
float startX = mp_obj_get_float_to_f(args[0]);
```

(All other uses of `double` should be changed accordingly as well)

**Reply by laurensvalk** [reply]
And to explain: 

> undefined reference to atan

You can use `atanf`.

(Although you'd probably want to use `atan2f` since dividing `dy` is problematic when `dy = startY - pointY = 0`.)

**Reply by Redstonetill** [reply]
Thanks for the quick answer. Using `atanf` + `fmodf` and only float variables (single-precision) (`mp_obj_get_float_to_f`) seems to work just fine

**Reply by Redstonetill** [reply]
Fyi `dy` can never be zero since I am checking if the startY and pointY is the same, but still thanks for the tipp :)

**Reply by laurensvalk** [reply]
Glad it worked!

> Fyi ``dy`` can never be zero since I am checking if the startY and pointY is the same, but still thanks for the tipp :)

Checking two doubles or floats for equality as `startY == pointY` is generally not reliable. They may be almost arbitrarily close which still means dividing by zero. It is safer to test something like `abs(startY - pointY) < some_small_value`
